pip/pip-299.md

```diff
@@ -0,0 +1,83 @@
+
+# Motivation
+
+Since the frequent persistence of `Individual Deleted Messages` will magnify the amount of BK Written and increase the latency of ack-response, the Broker does not immediately persist it when receiving a consumer's acknowledgment but persists it regularly. 
+
+The data of cursor metadata is recommended to be less than 5MB; if a subscription's `Individual Deleted Messages` data is too large to persist, as the program grows for a long time, there will be more and more non-persistent data. Eventually, there will be an unacceptable amount of repeated consumption of messages when the Broker restarts.
+
+# Goal
+
+## In Scope
+
+To avoid repeated consumption due to the cursor metadata being too large to persist.
+
+## Out of Scope
+
+This proposal will not care about this scenario: if so many producers make the metadata of cursor `pulsar.dedup` cannot persist, the task `Take Deduplication Snapshot` will be in vain due to the inability to persist.
+
+# High-Level Design
+
+Provide a new config named `dispatcherPauseOnAckPersistentStateEnabled` for a new feature: stop dispatch messages to clients when reaching the limitation `managedLedgerMaxUnackedRangesToPersist`.
+- If the user does not care about that Individual Deleted Messages can not be fully persistent, resulting in a large number of repeated message consumption, then it can be set to `false`.
+- If the user cares about repeated consumption, at can accept a decline in consumption speed when cursor metadata is too large to persist, it can be set to `true`.
+
+
+# Detailed Design
+### Public API
+
+**broker.conf**
+```
+/**
+ * If true, the dispatcher will pause the delivery of messages to clients if the cursor metadata(including ack state) reaches ${managedLedgerMaxAckStateInBytesToPersist}.
+ * The default value is "true".
+ * Note: If you set it to "false," when cursor metadata reaches ${managedLedgerMaxAckStateInBytesToPersist}, it can not persist, which might lead to repeated consumption.
+ */
+boolean dispatcherPauseOnAckPersistentStateEnabled;
+
+```
+
+## Design & Implementation Details
+
+Cache the range count of the Individual Deleted Messages in the memory when doing persist cursor metadata to BK. Stuck delivery messages to clients if reaching the limitation `managedLedgerMaxUnackedRangesToPersist`.
+
+Since the cache will not be updated in time, the actual count will decrease when clients acknowledge messages(but it does not persist immediately, so the cached value does not update immediately), the cached value is an estimated value.
+
+# Metrics
+
+**pulsar_cursor_meta_too_large_to_persist_state**
+- Description:
+    - 0: The cursor metadata size is fine.
+    - 1: The cursor metadata is too large to persist.
+- type: Enumeration
+- labels: `[cluster, namespace, topic, subscription]`
+
+# Alert
+- If `pulsar_cursor_meta_too_large_to_persist_state` is `1` for a long time, it indicates that the client might be stuck due to something wrong.
```

Comment:
Hmm, I don't think we need a new metric. If the message dispatch is stopped due to the acknowledgment state persistence. The backlog will continue to increase. Users usually only need to monitor the backlog.

Instead, I think we should add a new field to topic stats like `blockedSubscriptionOnUnackedMsgs`. 
Maybe we can use `blockedSubscriptionOnAckStatePersistent`

Reply:
Removed the metrics and added a subscription stat `isBlockedSubscriptionOnAckStatePersistent`